### PR TITLE
Compiler-support highlighting

### PR DESCRIPTION
**Pull request by david-christiansen**
This creates Agda-style highlighting. Note that it's also an incompatibility in the protocol, and clients will need to update to support the new Idris.

See https://github.com/idris-lang/Idris-dev/pull/2250 for details.

### CODE REVIEW

**Comment by david-christiansen** [comment]
The test failure was due to the use of a feature from Emacs 24.4. `idris-compat` has been updated to define it if it's missing.

**Comment by david-christiansen** [comment]
FYI: this version is backwards-compatible with older versions of Idris.

File: idris-highlight-input.el [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+(defun idris-highlight-column (idris-col)
+  "Compute the Emacs position offset of the Idris column IDRIS-COL, for highlighting.
+
+In particular, this takes bird tracks into account in literate Idris."
+  (+ idris-col (if (idris-lidr-p) 1 -1)))
+
+(defun idris-highlight-input-region (buffer start-line start-col end-line end-col highlight)
+  "Highight in BUFFER using an overlay from START-LINE and START-COL to END-LINE and END-COL and the semantic properties specified in HIGHLIGHT."
+  (when idris-semantic-source-highlighting
+    (if (not (or (and (> end-line start-line))
```
Comment (hannesm) [inline idris-highlight-input.el]: is this `and` with a single argument here?

Comment (hannesm) [inline idris-highlight-input.el]: can we please invert the conditional? this `not` and `or` and `and` make it hard to comprehend (at least in my head)

File: idris-prover.el [hunk 2]
```diff
@@ -218,7 +218,7 @@ left margin."
           (let ((tactic-cmd (replace-regexp-in-string
                              "\\`[ \t\n]*" ""
                              (replace-regexp-in-string "" " " tactic-text))))
-            (idris-rex () (list ':interpret tactic-cmd)
+            (idris-rex () (list ':interpret tactic-cmd) nil
```
Comment (hannesm) [inline idris-prover.el]: what is this change doing? and why are there 2 replace-regexp-in-string?

Comment (david-christiansen) [inline idris-prover.el]: I don't remember actually changing that line - my guess is that it's a Git hiccup related to the `^M` literal, which I can replace with something more readable. The reason for two replace-regexp-in-string calls is that the replacement string is different. The first removes leading whitespace, the second fixes up odd Windows newlines in parsed tactics, which the underlying Idris prover doesn't understand.

**Comment by hannesm** [comment]
thanks!

**Pull request merged**